autogpt/core/task/agnostic_task.py

```diff
@@ -0,0 +1,45 @@
+from autogpt.agent import Agent
+from autogpt.config import AIConfig, Config
+from autogpt.core.task.base import Task
+from autogpt.memory.vector import get_memory
+from autogpt.prompts.prompt import DEFAULT_TRIGGERING_PROMPT
+from autogpt.workspace import Workspace
+from tests.integration.agent_factory import get_command_registry
+
+
+class AgnosticTask(Task):
+    user_input: str
+
+    def run(self) -> None:
+        config = Config()
+        config.set_continuous_mode(False)
+        config.set_temperature(0)
+        config.plain_output = True
+        command_registry = get_command_registry(config)
+        config.memory_backend = "no_memory"
+        workspace_directory = Workspace.get_workspace_directory(config)
+        workspace_directory_path = Workspace.make_workspace(workspace_directory)
+
+        Workspace.build_file_logger_path(config, workspace_directory_path)
+        ai_config = AIConfig(
+            ai_name="Auto-GPT",
+            ai_role="a multi-purpose AI assistant.",
+            ai_goals=[self.user_input],
+        )
+
+        ai_config.command_registry = command_registry
+        system_prompt = ai_config.construct_full_prompt(config)
+
+        agent = Agent(
+            ai_name="Auto-GPT",
+            memory=get_memory(config),
+            command_registry=command_registry,
+            ai_config=ai_config,
+            config=config,
+            next_action_count=0,
+            system_prompt=system_prompt,
+            triggering_prompt=DEFAULT_TRIGGERING_PROMPT,
+            workspace_directory=str(workspace_directory_path),
+        )
+
+        agent.start_interaction_loop()
```

Comment:
this is just v1, this will be cleaner, there are other things to do to make it cleaner.

Reply:
I think you want a task to _just_ be data.  I.e. if we think about the set of things we need it looks like:

```python
import abc
from pydantic import BaseModel

class Task(BaseModel):
    """Jsonifiable representation of a task"""
    user_input: str

class Deliverable(BaseModel):  # Or artifact, or whatever you want to call this
    """"Jsonifiable representation of an agent output"""
    path: str

class Evaluation(BaseModel):
    """Jsonifiable representation of an evaluation of an agent against a benchmark task"""
    # Whatever you guys have been thinking about for evaluation results here
    success: bool

class PerformanceMetrics(BaseModel):
    """Jsonifiable representation of the resource performance of an agent on a benchmark task."""
    runtime: float
    # For later, since this is non-trivial to introspect
    tokens_used: int = 0
    cycle_count: int = 0
    api_cost: float = 0.0
   ...etc...


class Challenge(abc.ABC, BaseModel):
    task: Task
    deliverables: list[Deliverable]

    @abc.abstractmethod
    def evaluate_deliverables(self, deliverables: list[Deliverable]) -> Evaluation:
        ...
        
```

Then have a harness like:

```python 


def run_challenge(challenge: Challenge) -> tuple[Evaluation, PerformanceMetrics]:
    task = challenge.task
    start = time.time()
    agent = bootstrap_agent(task.user_input)
    agent.run_interaction_loop()
    end = time.time()
    
    deliverables = []
    for deliverable in challege.deliverables:
        try:
            agent.workspace.get_path(deliverable.path)
        except:
            deliverables.append(Deliverable(path='no result produced')

    evaluation = challenge.evaluate_deliverables(deliverables)
    performance_metrics = PerformanceMetrics(runtime=end-start)
   
    return evaluation, performance_metrics
   
def bootstrap_agent(user_input: str) -> Agent:
    config = Config()
    config.set_continuous_mode(False)
    config.set_temperature(0)
    config.plain_output = True
    command_registry = get_command_registry(config)
    config.memory_backend = "no_memory"
    workspace_directory = Workspace.get_workspace_directory(config)
    workspace_directory_path = Workspace.make_workspace(workspace_directory)

    Workspace.build_file_logger_path(config, workspace_directory_path)
    ai_config = AIConfig(
        ai_name="Auto-GPT",
        ai_role="a multi-purpose AI assistant.",
        ai_goals=[self.user_input],
    )

    ai_config.command_registry = command_registry
    system_prompt = ai_config.construct_full_prompt(config)

    agent = Agent(
        ai_name="Auto-GPT",
        memory=get_memory(config),
        command_registry=command_registry,
        ai_config=ai_config,
        config=config,
        next_action_count=0,
        system_prompt=system_prompt,
        triggering_prompt=DEFAULT_TRIGGERING_PROMPT,
        workspace_directory=str(workspace_directory_path),
    )

    return agent
```

Then you could literally have all your challenges just defined somewhere in a json file.